### ISSUE TITLE
[Bug]: occ files:scan-app-data consumes too much memory

### ISSUE THREAD

**Issue by Derkades** [nextcloud/server#40919]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

When running occ files:scan-app-data, the memory used by occ quickly rises until it crashes when the 4GB memory limit is exhausted:
```
[14-Oct-2023 19:54:06 UTC] PHP Fatal error:  Allowed memory size of 4294967296 bytes exhausted (tried to allocate 16777224 bytes) in /var/www/html/lib/private/Files/Cache/Cache.php on line 602
```

I would not expect a file scanner to use so much memory. My Nextcloud data directory contains 235k files, of which 17k in appdata and 100k in updater.

### Steps to reproduce

1. Run `occ files:scan-app-data`

### Expected behavior

The memory usage of the file scanner should probably not increase with the number of files

### Installation method

Community Manual installation with Archive

### Nextcloud Server version

27

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Upgraded to a MAJOR version (ex. 22 to 23)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

**Comment by joshtrichards** [comment]
When running this on one of my instances where the `appdata_*` folder has far ~4x the number of files as yours memory never goes that high.

`/var/www/html/lib/private/Files/Cache/Cache.php on line 602` looks to be where things are getting removed from the `filecache` not added. Maybe the issue is really the opposite: there were previously lots more files in one of your subfolders within `appdata_` perhaps. Hmm.

If you run it in verbose mode does it happen on any particular file/folder?

I suppose an event could also trigger it. 

Also, out of curiosity, for what purpose are you manually scanning your app data folder?

**Comment by Derkades** [comment]
Indeed, I deleted all previews which is why I wanted to rescan app data. I ended up "resolving" the issue by deleting all preview files from `oc_filecache` manually and then running the command again.

**Comment by solracsf** [comment]
PR at https://github.com/nextcloud/server/pull/41272

**Comment by Derkades** [comment]
Thanks!
